File: engine/model/src/main/scala/net/server/guild/MapleAlliance.scala
package net.server.guild

import scala.jdk.CollectionConverters._

class MapleAlliance(var name: String, var id: Int) {
  private var _guilds: List[Int] = List()

  var notice: String = ""

  var capacity: Int = 0

  var rankTitles: Array[String] = Array("Master", "Jr. Master", "Member", "Member", "Member")

  def removeGuild(guildId: Int): Unit = {
    _guilds.synchronized {
      _guilds = _guilds.filter(id => id != guildId)
    }
  }

  def addGuild(guildId: Int): Unit = {
    _guilds.synchronized {
      if (_guilds.size == capacity || _guilds.contains(guildId)) {
        return
      }
      _guilds = _guilds.appended(guildId)
    }
  }

  def rankTitle(rank: Int): String = {
    rankTitles(rank - 1)
  }

  def guilds: java.util.List[Integer] = {
    _guilds.map(i => i: java.lang.Integer).asJava
  }

  def increaseCapacity(increase: Int): Unit = capacity += increase
}

File: engine/engine-model/src/main/java/net/server/guild/MapleAlliance.java
package net.server.guild;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public record MapleAlliance(String name, Integer id, Integer capacity, String notice, String[] rankTitles,
                            List<Integer> guilds) {
   public MapleAlliance(String name, Integer id) {
      this(name, id, 0);
   }

   public MapleAlliance(String name, Integer id, Integer capacity) {
      this(name, id, capacity, "", new String[]{"Master", "Jr. Master", "Master", "Master", "Master"});
   }

   public MapleAlliance(String name, Integer id, Integer capacity, String notice, String[] rankTitles) {
      this(name, id, capacity, notice, rankTitles, Collections.emptyList());
   }

   public MapleAlliance setRankTitles(String[] rankTitles) {
      return new MapleAlliance(name, id, capacity, notice, rankTitles, guilds);
   }

   public MapleAlliance setGuilds(List<Integer> guilds) {
      return new MapleAlliance(name, id, capacity, notice, rankTitles, guilds);
   }

   public MapleAlliance addGuild(Integer guildId) {
      if (guilds.size() == capacity || guilds.contains(guildId)) {
         return this;
      }

      List<Integer> newGuilds = new ArrayList<>(guilds);
      newGuilds.add(guildId);
      return new MapleAlliance(name, id, capacity, notice, rankTitles, Collections.unmodifiableList(newGuilds));
   }

   public MapleAlliance removeGuild(Integer guildId) {
      return new MapleAlliance(name, id, capacity, notice, rankTitles, guilds.stream()
            .filter(id -> !id.equals(guildId))
            .collect(Collectors.toUnmodifiableList()));
   }

   public MapleAlliance increaseCapacity(Integer increase) {
      return new MapleAlliance(name, id, capacity + increase, notice, rankTitles, guilds);
   }

   public MapleAlliance setNotice(String notice) {
      return new MapleAlliance(this.name, id, capacity, notice, rankTitles, guilds);
   }

   public String rankTitle(Integer rank) {
      if (rank < 1 || rank > rankTitles.length) {
         throw new UnsupportedOperationException();
      }
      return rankTitles[rank - 1];
   }
}
